proposals/NNNN-module-aliasing-for-disambiguation.md

```diff
@@ -0,0 +1,226 @@
+```
+
+In the following steps, we will use a compiler invocation command to demonstrate how the conflicting modules can be aliased. 
+
+1. First, take the Utils module from swift-game, and rename it to be unique; let’s call it GameUtils. We will need to compile the module (a) by giving a new name GameUtils while (b) treating any references to Utils in its source files as GameUtils.  
+    a. The first part (renaming) can be achieved via passing the new name (GameUtils) to `-module-name` and an output path flag, `-o`,  `-emit-module-path`, or `-emit-module-interface-path`, which are all existing flags. The resulting binary then, for example, will be `/path/to/GameUtils.swiftmodule` (instead of `Utils.swiftmodule`).
+    b. The second part (treating references to Utils in source files as GameUtils) will be addressed by introducing a new compiler flag `-module-alias [name]=[binary_name]`, where the name is the module name that appears in source files (Utils), while the binary_name is the name of the .swiftmodule (GameUtils), the canonical name used for file contents including symbol mangling. Combined with step (a), the compiler invocation command will be `swiftc -module-name GameUtils -emit-module-path /path/to/GameUtils.swiftmodule -module-alias Utils=GameUtils ...`.    
+2. Similarly, we can rename Utils from swift-draw as DrawUtils, following the step 1.  
+3. Finally, we build App (without the aliasing flag) and directly import GameUtils and/or DrawUtils. 
+
+```
+[App]
+
+import GameUtils
+import DrawUtils
+```
+In practice, a dependency graph will be more complex than the scenario described above. As an example, we have the following modified scenario. App imports module Game, which imports module Utils from the same package. App also imports another module called Utils from a different package.  
+
+```
+App 
+|— Module Game (from package ‘swift-game’)
+|— Module Utils (from package ‘swift-game’)
+|— Module Utils (from package ‘swift-draw’)
+```
+
+```
+[Module Game] // swift-game
+
+import Utils  // swift-game
+public func start(level: Utils.Level) { ... }
+```
+
+As with the first scenario, the Utils modules are conflicting, so we need to perform the steps 1-2 above. Then we need to build module Game by applying `-module-alias Utils=GameUtils`, so that the references to Utils in source files of module Game are compiled as GameUtils without making any source changes. The compiler invocation command to build Game then is `swiftc -module-name Game -module-alias Utils=GameUtils ...`. App can then import module Game and any of the renamed Utils as needed. 
+
+As seen above, module aliasing can be done directly via a compiler invocation command, but most users do not interact with the command directly. Thus we plan to provide an easier access via new build configs which can be adopted by any build systems; in particular, we will focus on how it can be adopted via SwiftPM, described in a later section. 
```

Comment:
```suggestion
As seen above, module aliasing can be done directly via compiler invocation flags. However, most users use high-level build systems rather than interacting with the command line directly. We will describe how this feature can be adopted in SwiftPM in a later section. 
```

I'd leave the part about "adopted by any build system" out; I don't see anything like that in the proposal, and I don't know how it would work.  You're adding command line flags that any build system can use, and you're adding features to SwiftPM that will translate to these flags.

I'm not sure I agree with this ordering of the presentation, though.  If you really think that the SwiftPM features are the way most users will interact with this, then I think you should describe those as the Proposed Solution, and the frontend flags should be part of the Detailed Design.  Or maybe Proposed Solution first describes it at a high level without talking about either, then talk about the SwiftPM flags, then Detailed Design goes into more detail about both.  That is, the flow would be:

> **Proposed Solution**
>
> We think module aliases solve the problem.  Swift already lets you set the name of a module.  Module aliases let you also tell Swift that a module name is actually an alias for a different module.  Together, these low-level tools allow conflicts to be resolved by giving modules a unique name while using aliases to avoid the need to change any source code.  All of this can be done directly from SwiftPM.
>
> Here is an example: ...
>
> We need to rename these modules when building them.  App will then use them under the renamed names: ...
>
> Here's how to do that in SwiftPM: ...
>
> That example was uncommon.  Here's a more likely one with an indirect dependency: ...
>
> We need to rename one of these modules, both internally and in the modules that use it: ...
>
> Here's how to do that in SwiftPM: ...
>
> **Detailed Design**
>
> Command-line flags
>
> SwiftPM features